### ISSUE TITLE
Bad performance of DefaultExecutionPlan.hasOverlap

### ISSUE THREAD

**Issue by tobiasbaum** [gradle/gradle#11792]
Gradle seemed to be doing nothing for quite a while in some of my builds. Profiling showed that it spends a lot of time in org.gradle.execution.plan.DefaultExecutionPlan.hasOverlap(). Looking at the code, it should not be too hard to optimize the performance.

### Expected Behavior
Gradle should be fast.

### Current Behavior
In some builds, Gradle spends more than a minute in the given function. No tasks are running at that time. Here's a screenshot of the profiling result: 
![gradlePerformanceIssueHasOverlap](https://user-images.githubusercontent.com/4025550/71720435-06702500-2e22-11ea-98c8-8e4711f63da3.png)

### Context
We have a large (albeit not monstruos) project: ~20,000 Java files, > 2,000,000 LOC, ~300 sub-projects, several thousand tasks overall, of which 670 were relevant for the respective build. We try to keep build times to a minimum, especially for incremental builds.

### Steps to Reproduce 
I noticed the behavior several times but did not manage to reproduce the behavior consistently. But I don't think exact reproduction is needed, as the code can be optimized for all use cases.

### Your Environment
We're on Gradle 6.0.1. The given build scan is from the same project and target, but not from the same build in which I noticed the problem and did the profiling.
Build scan URL: https://gradle.com/s/x7npzoe7k3orm

### Proposed solution
Currently, the algorithm runs in quadratic time when both parameters (paths1 and paths2) are large. By using a TreeSet as a helper data structure and its ceiling() and floor() methods, it should be possible to make it run in log-linear time. If help is needed, I could try to create a pull request.

**Comment by wolfs** [comment]
The performance of `hasOverlap`, though it probably doesn't matter much. As you saw in your micro-benchmarks, even for medium sized paths to check for overlap, the current algorithm is pretty fast.

If you have 10 tasks running in parallel (which is already quite a few), and you want to schedule a new task, then the expectation is that there isn't an overlap of outputs with any of the running task. So `hasOverlap` would be called only once per task which is ready to be executed. The sizes of both sets would also be pretty small, since a task probably has 1-2 output locations.

I wonder why you see a performance problem with that method in your build. Do you have a task with many outputs or a task that destroys many locations (aka `clean`)?

I ran the performance tests for your branch, and didn't see an improvement for most of our test builds. I would have expected our `largeJavaMultiproject` build to show up, given that has 500 submodules.

Using your PR, can you measure a significant speed-up of your build? You can use [Gradle profiler](https://github.com/gradle/gradle-profiler) to measure.

**Comment by tobiasbaum** [comment]
I ported my PR back to 6.0.1 and put it into our production build about a week ago. It did improve the build time (if I remember correctly it was something like 1:30min -> 40s on a build where many tasks were up-to-date or cached) and the "gradle seems to do nothing with all workers IDLE" phases have become noticeably shorter. Given that I observed a 2s speedup in my microbenchmarks, having a similar speedup for something like 20 calls could explain the improvement.

From my further observations, it could be that the problem has to do with having large "node_modules" directories for Angular projects as task inputs for several of the subprojects. At least, the slowness often occurred when theses tasks were around.

On the stats: In my local build, I usually have 12 tasks running in parallel (on a 16 core machine) and I guess that quite a number of tasks are often ready to run.

I'm not that surprised that you haven't noticed a problem/improvement in your existing tests, otherwise you would have probably fixed it before ;)

I haven't found time to look into the class you mentioned in the PR, but I hope that I'll come around to test it tomorrow.

**Comment by tobiasbaum** [comment]
Here are two profile reports, one with optimization and one without. I tried to ensure that cache, daemon etc are warmed up and also did some more repetitions, with similar results. The improvement is even larger than what I had written above, is going down from 2m18.87s to 36.016s

[profiles.zip](https://github.com/gradle/gradle/files/4076472/profiles.zip)

**Comment by foto-andreas** [comment]
Hi,

the PR is still not merged. Why? Another Gradle release I have to patch before using in production.

Andreas

**Comment by wolfs** [comment]
@tobiasbaum @nemeth-z Which task is causing this problem? I think the actual issue is not that `hasOverlap` is not very performant with many files, but that many files are checked for overlap in the first place.

I think the reason for this is that we add all files in the output file collections here: https://github.com/gradle/gradle/blob/2862ff1dd6809f9399d6333893469004c98d1689/subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNode.java#L202-L217

For a file tree, that means walking the whole file system for the output. We should instead add only the root of the file collection there - for example the root directory of the file tree. This will reduce the number of files in the set and avoids walking the file system when scheduling tasks.

Can you confirm that the task which takes very long in checking `hasOverlap` either has
- A file tree as a destroyable, e.g. a `Delete` task deleting a `fileTree`
- A file tree as output files?

**Comment by nemeth-z** [comment]
I can confirm the second one in case of the Node plugin. However the maintainer has fixed it by changing it to an output directory instead, so it does not affect me any longer (at least for now).

**Comment by tobiasbaum** [comment]
I'm also pretty sure that we have the second case, perhaps also the first.

**Comment by wolfs** [comment]
I published a nightly from https://github.com/gradle/gradle/pull/13066: `6.5-20200509104259+0000`

This should fix the performance problems you experience, since it removes the root cause: having to check overlaps between a large number of files in the first place.

Can everyone affected by this issue try out this nightly version and confirm that it fixes the problem for them?

**Comment by tobiasbaum** [comment]
I did a performance test with the nightly. Median results:

- My optimized 6.1: 1 m 10 s
- Vanilla 6.4: 4 m 13 s
- Your nightly 6.5: 1 m 3 s

So your fix seems to work, too (and even a bit better) :) Thanks!

**Comment by wolfs** [comment]
Awesome, thank you @tobiasbaum for trying it out. The fix will be in 6.5.
